text/0000-dyn-error-generic-member-access.md

```diff
@@ -0,0 +1,356 @@
+And finally, we create an error reporter that prints the error and its source
+recursively along with the location data if it was gathered.
+
+```rust
+struct ErrorReporter(Box<dyn Error + Send + Sync + 'static>);
+
+impl fmt::Debug for ErrorReporter {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let mut current_error = Some(self.0.as_ref());
+        let mut ind = 0;
+
+        while let Some(error) = current_error {
+            writeln!(fmt, "    {}: {}", ind, error)?;
+
+            if let Some(location) = error.context::<Location>() {
+                writeln!(fmt, "        at {}:{}", location.file, location.line)?;
+            }
+
+            ind += 1;
+            current_error = error.source();
+        }
+
+        Ok(())
+    }
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+There are two additions necessary to the standard library to implement this
+proposal:
+
+Add a function for dyn Error trait objects that will be used by error
+reporters to access members given a generic type. This function circumvents
+restrictions on generics in trait functions by being implemented for trait
+objects only, rather than as a member of the trait itself.
+
+```rust
+impl dyn Error {
+    pub fn context<T: Any>(&self) -> Option<&T> {
+        self.provide_context(TypeId::of::<T>())?.downcast_ref::<T>()
+    }
+}
+```
+
+With the expected usage:
+
+```rust
+// With explicit parameter passing
+let spantrace = error.context::<SpanTrace>();
+
+// With a type inference
+fn get_spantrace(error: &(dyn Error + 'static)) -> Option<&SpanTrace> {
+    error.context()
+}
+```
+
+Add a member to the `Error` trait to provide the `&dyn Any` trait objects to
+the `context` fn for each member based on the type_id.
+
+```rust
+trait Error {
+    /// ...
+
+    fn provide_context(&self, id: TypeId) -> Option<&dyn Any> {
+        None
+    }
+}
+```
+
+With the expected usage:
+
+```rust
+fn provide_context(&self, type_id: TypeId) -> Option<&dyn Any> {
+    if id == TypeId::of::<Location>() {
+        Some(&self.location)
+    } else {
+        None
+    }
+}
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+* The API for defining how to return types is cumbersome and possibly not
+  accessible for new rust users.
+    * If the type is stored in an Option getting it converted to an `&Any` will
+      probably challenge new devs, this can be made easier with documented
+      examples covering common use cases and macros like `thiserror`.
+```rust
+} else if typeid == TypeId::of::<SpanTrace>() {
+    self.span_trace.as_ref().map(|s| s as &dyn Any)
+}
+```
+* When you return the wrong type and the downcast fails you get `None` rather
+  than a compiler error guiding you to the right return type, which can make it
+  challenging to debug mismatches between the type you return and the type you
+  use to check against the type_id
+    * The downcast could be changed to panic when it fails
```

Comment:
I'd call this an unknown rather than a drawback "how should the `context` fn handle failed downcasts?"